ocsp/responder/redis/checked_redis_source_test.go

```diff
@@ -0,0 +1,184 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/go-gorp/gorp/v3"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/db"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/ocsp/responder"
+	ocsp_test "github.com/letsencrypt/boulder/ocsp/test"
+	"github.com/letsencrypt/boulder/test"
+	"golang.org/x/crypto/ocsp"
+)
+
+// echoSource implements rocspSourceInterface, returning the provided response
+// and panicking if signAndSave is called.
+type echoSource struct {
+	resp *ocsp.Response
+}
+
+func (es echoSource) Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error) {
+	return &responder.Response{Response: es.resp, Raw: es.resp.Raw}, nil
+}
+
+func (es echoSource) signAndSave(ctx context.Context, req *ocsp.Request, cause string) (*responder.Response, error) {
+	panic("should not happen")
+}
+
+// recordingEchoSource acts like echoSource, but instead of panicking on signAndSave,
+// it records the serial number it was called with and returns the given secondResp.
+type recordingEchoSource struct {
+	echoSource
+	secondResp *responder.Response
+	ch         chan string
+}
+
+func (res recordingEchoSource) signAndSave(ctx context.Context, req *ocsp.Request, cause string) (*responder.Response, error) {
+	res.ch <- req.SerialNumber.String()
+	return res.secondResp, nil
+}
+
+// errorSource implements rocspSourceInterface, and always returns an error.
+type errorSource struct{}
+
+func (es errorSource) Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error) {
+	return nil, errors.New("sad trombone")
+}
+
+func (es errorSource) signAndSave(ctx context.Context, req *ocsp.Request, cause string) (*responder.Response, error) {
+	panic("should not happen")
+}
+
+// echoSelector always returns the given certificateStatus.
+type echoSelector struct {
+	db.MockSqlExecutor
+	status core.CertificateStatus
+}
+
+func (s echoSelector) WithContext(context.Context) gorp.SqlExecutor {
+	return s
+}
+
+func (s echoSelector) SelectOne(output interface{}, _ string, _ ...interface{}) error {
+	outputPtr, ok := output.(*core.CertificateStatus)
+	if !ok {
+		return fmt.Errorf("incorrect output type %T", output)
+	}
+	*outputPtr = s.status
+	return nil
+}
+
+// errorSelector always returns an error.
+type errorSelector struct {
+	db.MockSqlExecutor
+}
+
+func (s errorSelector) SelectOne(_ interface{}, _ string, _ ...interface{}) error {
+	return errors.New("oops")
+}
+
+func (s errorSelector) WithContext(context.Context) gorp.SqlExecutor {
+	return s
+}
+
+func TestCheckedRedisSourceSuccess(t *testing.T) {
+	serial := big.NewInt(17777)
```

Comment:
Having a different serial for each of these is just for funtimes, right? I don't think I see anything that would fall over if two of these test cases were using the same serial...

Reply:
Yes, just for funtimes; and on the off chance that a serial number shows up in the test output. Perhaps it would be clearer to use the same, boring serial number for all test cases. I could go either way in the future. :-)